client/jetpack-connect/controller.js

```diff
@@ -184,6 +178,12 @@ export function connect( context, next ) {
 export function signupForm( context, next ) {
 	analytics.pageView.record( 'jetpack/connect/authorize', 'Jetpack Authorize' );
 
+	const isLoggedIn = !! getCurrentUserId( context.store.getState() );
+	if ( retrieveMobileRedirect() && ! isLoggedIn ) {
+		// Force login for mobile app flow. App will this request and prompt native login.
+		return window.location.replace( login( { isNative: true, redirectTo: context.path } ) );
```

Comment:
Comment lost in rebase https://github.com/Automattic/wp-calypso/pull/21729#pullrequestreview-90529087

> Now making an explicit server request, which is easier for mobile app to intercept. Request will never actually hit the server.